pkg/manifest/manifest.go

```diff
@@ -99,22 +148,77 @@ func ManifestsFromFiles(files []string) ([]Manifest, error) {
 }
 
 // ParseManifests parses a YAML or JSON document that may contain one or more
-// kubernetes resources.
-func ParseManifests(r io.Reader) ([]Manifest, error) {
+// kubernetes resources. Parsing and structure related errors are returned in
+// the first error parameter and cause an immediate return. Duplicate resource
+// errors are returned in the error list. Processing continues after a
+// duplicate resource error allowing all duplicates to be identified in the
+// given input.
+func ParseManifests(r io.Reader) ([]Manifest, error, []error) {
+	theseIds := make(map[ResourceId]bool)
+	var dupErrors []error
 	d := yaml.NewYAMLOrJSONDecoder(r, 1024)
 	var manifests []Manifest
 	for {
 		m := Manifest{}
 		if err := d.Decode(&m); err != nil {
 			if err == io.EOF {
-				return manifests, nil
+				return manifests, nil, dupErrors
 			}
-			return manifests, errors.Wrapf(err, "error parsing")
+			return manifests, errors.Wrapf(err, "error parsing"), dupErrors
 		}
 		m.Raw = bytes.TrimSpace(m.Raw)
 		if len(m.Raw) == 0 || bytes.Equal(m.Raw, []byte("null")) {
 			continue
 		}
+		if resourceId, err := unmarshalResourceId(m.Raw); err != nil {
+			return manifests, err, dupErrors
+		} else {
+			m.resourceId = resourceId
+		}
+		if err := addIfNotDuplicateResource(m, theseIds); err != nil {
+			dupErrors = append(dupErrors, err)
+		}
 		manifests = append(manifests, m)
 	}
 }
+
+func unmarshalResourceId(manifestBytes []byte) (ResourceId, error) {
+	var resourceId ResourceId
+	var yamlId ResourceIdYaml
+	if err := goyaml.Unmarshal(manifestBytes, &yamlId); err != nil {
+		return resourceId, errors.Wrapf(err, "unable to unmarshall resource id fields")
+	}
+	if !validReourceId(yamlId) {
+		return resourceId, errors.New("Kubernetes objects must contain required fields apiVersion, kind, and metadata.name")
```

Comment:
It's probably more sustainable and easier to debug to use `validateResourceId(id ResourceIdYaml) error` that complains directly about the missing property.  Also, might be worth linking [these docs][1], either in the error message or a Go comment, to back up the requirement.

[1]: https://github.com/kubernetes/community/blob/2f26b93380f504e4c11effdae1eee7bbe6648d5b/contributors/devel/sig-architecture/api-conventions.md#resources

Reply:
I added a method comment with a kube link that I thought explained it well. And since Group could be empty I left it off. Also turns out that the existing unmarshall captures `kind` related issues but not an empty `name`.